modules/custom/wim_cvdr/wim_cvdr.module

```diff
@@ -0,0 +1,252 @@
+
+  // Use caching to improve performance.
+  if ($cached = cache_get('wim_cvdr_items')) {
+    $items = $cached->data;
+  }
+  else {
+    $xml = _wim_cvdr_read_xml();
+    if ($xml) {
+      $items = _wim_cvdr_process_xml($xml);
+    }
+  }
+
+  return $items;
+}
+
+/**
+ * Request the XML from the webservice.
+ */
+function _wim_cvdr_read_xml() {
+  $xml_url = variable_get('wim_cvdr_url');
+
+  // Request the XML from the webservice.
+  $xml_response = drupal_http_request($xml_url, ['timeout' => 10]);
+
+  if ($xml_response->code !== '200' || strpos($xml_response->headers['content-type'], 'text/xml') === FALSE) {
+    // Log error and quit.
+    watchdog('WIM CVDR',
+      'Could not retrieve XML. Debug: !array',
+      ['!array' => print_r($xml_response, TRUE)],
+      WATCHDOG_ERROR);
+    return FALSE;
+  }
+
+  // Return the XML data.
+  return $xml_response->data;
+}
+
+/**
+ * Process the XML by creating an array keyed on subject.
+ *
+ * @param string $xml
+ *   The xml in string format.
+ *
+ * @return array
+ *   With items keyed on subject and sorted.
+ */
+function _wim_cvdr_process_xml($xml) {
+  $items = [];
+  libxml_use_internal_errors(TRUE);
+  $xml_dom = new DOMDocument();
+  $xml_dom->loadXML($xml);
+
+  if ($xml === FALSE) {
+    // Log error and quit.
+    watchdog('WIM CVDR',
+      'XML is malformed and couldn\'t be properly loaded/parsed. Errors: !message',
+      ['!message' => print_r(libxml_get_errors(), TRUE)],
+      WATCHDOG_ERROR);
+    libxml_clear_errors();
+
+    return $items;
+  }
+
+  // Loop through the items in the XML.
+  foreach ($xml_dom->getElementsByTagName('gzd') as $item) {
+    // Retrieve the needed data by using xpath queries.
+    $xpath = new DOMXpath($xml_dom);
+    $url = $item->getElementsByTagName('publicatieurl_xhtml')->item(0);
+    $subject = $xpath->query('.//overheidrg:onderwerp', $item)->item(0);
+    $title = $xpath->query('.//dcterms:title', $item)->item(0);
+    $date = $xpath->query('.//overheidrg:uitwerkingtredingDatum', $item)
+      ->item(0);
+
+    // Only proceed if everything is available.
+    if (isset($url, $subject, $title, $date)) {
+      $url = $url->nodeValue;
+      $subject = ucfirst($subject->nodeValue);
+      $title = $title->nodeValue;
+      $date = $date->nodeValue;
+
+      // Check if no expire date is set or expire date is in the future.
+      if (empty($date) || strtotime($date) > REQUEST_TIME) {
+        // Fallback when no subject is set for this item.
+        if (empty($subject)) {
+          $subject = t('Other');
+        }
+
+        // Save the item keyed by its subject.
+        $items[$subject][] = [
+          'title' => $title,
+          'url' => $url,
+        ];
+      }
+    }
+  }
+
+  // Sort items on key (subject) alphabetically.
+  ksort($items);
+
+  // Save the cached results to improve performance.
+  $expire = REQUEST_TIME + (variable_get('wim_cvdr_cache_time', 60) * 60);
+  cache_set('wim_cvdr_items', $items, 'cache', $expire);
```

Comment:
Ideally I'd still like to move this to `_wim_cvdr_retrieve_elements` so that the `cache_set` and `cache_get` are close together which is easier for future developers. It also gives us future opportunities to circumvent the cache if this is required.